service-wfs3/src/main/java/org/oskari/service/wfs3/OskariWFS3Client.java

```diff
@@ -140,29 +145,49 @@ public static SimpleFeatureCollection getFeatures(OskariLayer layer,
 
                 validateResponse(conn, CONTENT_TYPE_GEOJSON);
                 geojson = readMap(conn);
-                sfc = GeoJSONReader2.toFeatureCollection(geojson, schema, transformCRS84ToTargetCRS);
+                sfc = GeoJSONReader2.toFeatureCollection(geojson, schema, transformCRS84ToTargetCRS, postFilter);
                 numFeatures += sfc.size();
                 pages.add(sfc);
                 next = getLinkHref(geojson, "next");
             }
+
+            if (pages.size() == 1) {
+                return pages.get(0);
+            }
+            return new PaginatedFeatureCollection(pages, schema, "FeatureCollection", hardLimit);
         } catch (IOException e) {
             throw new ServiceRuntimeException("IOException occured", e);
         } catch (MismatchedDimensionException | TransformException e) {
             throw new ServiceRuntimeException("Projection transformation failed", e);
         }
+    }
+
+    private static int clamp(int value, int min, int max) {
+        return value > max ? max : value < min ? min : value;
```

Comment:
Would be more lines but easier to read in my opinion:
```suggestion
        if (value > max) {
            return max;
        } else if (value < min) {
            return min;
        }
        return value;
```